### ISSUE TITLE
Traefik in from-scratch Docker container exits with exit code 0 if RBAC permissions on Kubernetes are insufficient

### ISSUE THREAD

**Issue by timoreimann** [containous/traefik#1785]
### Do you want to request a *feature* or report a *bug*?
Bug.

### What did you do?
Start Traefik on Kubernetes with RBAC permissions set to allow access to Ingresses, Services, and Endpoints. Specifically, I did not permit access to Secrets.

### What did you expect to see?
Traefik humming along nice and smooth.

### What did you see instead?
Traefik exits with error code 2 when it cannot access Secrets.

### Output of `traefik version`: (_What version of Traefik are you using?_)
latest master.

### If applicable, please paste the log output in debug mode (`--debug` switch)
Relevant logs:

```
E0625 21:28:59.568275       1 reflector.go:300] github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to watch *v1.Secret: the server does not allow access to the requested resource (get secrets)
log: exiting because of error: log: cannot create log: open /tmp/traefik.traefik-ingress-controller-wgt5n.unknownuser.log.ERROR.20170625-212859.1: no such file or directory
```

The last line is from glog and responsible for the actual premature program termination.

The issue is happening because of two events happening simultaneously:

- Traefik requires permissions to Secrets even when basic auth isn't used (#1784).
- client-go configures glog to log such access failures to a file but the required `/tmp` directory isn't available on a `FROM scratch` Docker container, leading to the eventual failure.

The proper fix is likely to make client-go not log to a file by default. In the meantime though, we can mitigate the problem by fixing #1784. In general, Traefik shouldn't stop running because of a permission problem with one provider.

For the Secrets case specifically, a workaround for now is to enable Secrets access on RBAC.

**Comment by dtomcej** [comment]
@timoreimann Are you suggesting a RBAC test on startup to verify permissions? Or are you wanting to have secrets access verified via API before requesting? Or are you suggesting to handle access failures better?

**Comment by timoreimann** [comment]
@dtomcej IMHO with Traefik already logging access failures on a log level higher than INFO, we've done our duty. It's on client-go to not try to log into a file by default (or, at a bare minimum, not have logrus exit if the log operation fails).

I am yet to verify the behavior with the latest client-go version. Until then, I think we should keep this issue open to track any upstream fix we may be able to incorporate in the (hopefully near) future.

**Comment by timoreimann** [comment]
I have even encountered a case now where Traefik/client-go/logrus terminated unrelated to an RBAC problem:

```
W0627 23:37:58.259690       1 reflector.go:319] github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/reflector.go:94: watch of *v1.Secret ended with: too old resource version: 174423 (175800)
log: exiting because of error: log: cannot create log: open /tmp/traefik.traefik-ingress-controller-3371602480-9pt0s.unknownuser.log.WARNING.20170627-233758.1: no such file or directory
```

Not sure how often it happens, but it's at least another failure vector.

**Comment by timoreimann** [comment]
I have been blaming logrus throughout this issue when in fact it was glog.

Update all of my comments accordingly.

**Comment by timoreimann** [comment]
Amended glog by a call to `debug.PrintStack()` to better see where the problematic calls are coming from:

```
E0702 23:41:29.843353       1 reflector.go:199] github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Endpoints: the server does not allow access to the requested resource (get endpoints)
goroutine 49 [running]:
runtime/debug.Stack(0xc42036c540, 0x60, 0x242)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x79
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x22
github.com/containous/traefik/vendor/github.com/golang/glog.create(0x1f79bcf, 0x5, 0xed0eb7ba9, 0xc43244e4c0, 0x2e8d400, 0x489b44, 0x486ce5, 0x1, 0x2, 0xc420344c80)
	/go/src/github.com/containous/traefik/vendor/github.com/golang/glog/glog_file.go:125 +0x35b
github.com/containous/traefik/vendor/github.com/golang/glog.(*syncBuffer).rotateFile(0xc42033d050, 0xed0eb7ba9, 0x3244e4c0, 0x2e8d400, 0x0, 0xec)
	/go/src/github.com/containous/traefik/vendor/github.com/golang/glog/glog.go:832 +0xa2
github.com/containous/traefik/vendor/github.com/golang/glog.(*loggingT).createFiles(0x2e8d760, 0xc400000002, 0xec, 0x13b)
	/go/src/github.com/containous/traefik/vendor/github.com/golang/glog/glog.go:867 +0xd8
github.com/containous/traefik/vendor/github.com/golang/glog.(*loggingT).output(0x2e8d760, 0xc400000002, 0xc4203406e0, 0x2d44827, 0xc, 0xc7, 0x0)
	/go/src/github.com/containous/traefik/vendor/github.com/golang/glog/glog.go:686 +0x4e2
github.com/containous/traefik/vendor/github.com/golang/glog.(*loggingT).printDepth(0x2e8d760, 0xc400000002, 0x2, 0xc420585e08, 0x1, 0x1)
	/go/src/github.com/containous/traefik/vendor/github.com/golang/glog/glog.go:646 +0x12d
github.com/containous/traefik/vendor/github.com/golang/glog.ErrorDepth(0x2, 0xc420585e08, 0x1, 0x1)
	/go/src/github.com/containous/traefik/vendor/github.com/golang/glog/glog.go:1106 +0x5d
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime.logError(0x2d64920, 0xc4202733d0)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:114 +0x75
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime.HandleError(0x2d64920, 0xc4202733d0)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:108 +0x69
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.(*Reflector).RunUntil.func1()
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/reflector.go:199 +0x5a
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait.JitterUntil.func1(0xc4204e3940)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:96 +0x5e
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait.JitterUntil(0xc4204e3940, 0x3b9aca00, 0x0, 0x100010001, 0xc420055020)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:97 +0xad
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait.Until(0xc4204e3940, 0x3b9aca00, 0xc420055020)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:52 +0x4d
created by github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.(*Reflector).RunUntil
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/reflector.go:201 +0x19d
E0702 23:41:29.843353       1 reflector.go:199] github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Endpoints: the server does not allow access to the requested resource (get endpoints)
log: exiting because of error: log: cannot create log: open /tmp/traefik.traefik-ingress-controller-3566636584-jw3z6.unknownuser.log.ERROR.20170702-234129.1: no such file or directory
```

**Comment by timoreimann** [comment]
After digging a bit deeper into the issue, I came to learn that passing the `-logtostderr` flag causes glog to pump out logs into STDERR instead of a file. I also verified that it works by setting the flag in the Kubernetes provider code like this:

```go
flag.Set("logtostderr", "true")
```

(Passing the flag on the CLI didn't work. My guess is that flaeg's structure-/tag-based approach prevents us from accepting any "leaking" flags.)

Apparently, even the Kubernetes infrastructure components set it [all over the place](https://github.com/kubernetes/kubernetes/search?l=Go&q=logtostderr&type=&utf8=%E2%9C%93).

There's an [open feature request](https://github.com/kubernetes/client-go/issues/18) to introduce a proper logging interface, but nothing has happened so far.

I'm afraid we'll have to go with the somewhat hackish solution for now.
